0023-keymanager-secret-sharing.md

```diff
@@ -0,0 +1,671 @@
+    security levels, etc.
+
+  - State re-encryption should be possible with keys derived from
+    a new shared secret.
+
+- Key manager restarts should not lose shares or committed data.
+
+### Protocol
+
+#### Offline
+
+- Assign non-zero ID numbers to all nodes and associate them with their
+  public keys.
+
+- The key manager owner selects a Schnorr group based on security
+  requirements and prepares the CHURP configuration and access policy.
+
+  - The Schnorr group can also be generated by one of the enclaves.
+    In this case, the owner only needs to define security parameters.
+
+#### Initialization
+
+- The key manager owner publishes the configuration in the consensus layer.
+
+- Key manager nodes interested in participating update their configuration
+  file and restart the node.
+
+  - To prevent the need for restarting the node, a command could be added
+    to the CLI.
+
+- Following the restart, the node's enclave prepares a non-zero-hole
+  verification matrix for the dealing phase and registers with its checksum
+  (public commitment).
+
+#### Dealing phase
+
+- Starts at the beginning of the first epoch with a sufficient number
+  of registered nodes.
+
+- The first t+2 nodes serve as dealers.
+
+  - Make sure to include enough nodes to prevent dealer corruption.
+
+  - Verification matrices of other nodes are disregarded, and their entropy
+    will not be included in the secret. However, they will still receive
+    a share.
+
+- The construction of the shared secret and dealing occurs offline through
+  a peer-to-peer network, following the specified enclave policy.
+
+- Each registered node (dealing):
+
+  - requests its bivariate shares (polynomials and non-zero-hole
+    verification matrices) from the dealers,
+
+  - validates received shares,
+
+  - verifies non-zero-hole verification matrices against the consensus layer,
+
+  - combines shares (adds polynomials and merges non-zero-hole verification
+    matrices),
+
+  - stores the result (full share) locally in the enclave's confidential
+    storage,
+
+  - sends a transaction containing the checksum of the merged matrix
+    to the consensus layer, confirming receipt of all shares.
+
+- If the timeout/epoch expires or checksums do not match, the registry
+  is cleared, and nodes must re-register.
+
+- Upon receiving confirmations from all registered nodes, the consensus layer
+  announces the new committee and begins collecting registrations for the
+  first handoff.
+
+  - Nodes can register for the handoff no earlier than one epoch in advance.
+
+- The dealers delete dealing data.
+
+- The committee starts serving requests.
+
+#### Serving
+
+- To construct a derived key, a key manager client must contact at least
+  the threshold number of nodes within the committee to acquire the necessary
+  derived key shares.
+
+- The committee responds exclusively to nodes defined in the policy.
+
+- Blame detection should/can be added later (expensive).
```

Comment:
Even if there is no blame assignment initially, is it still possible to detect that _some_ corruption occurred?

Reply:
I think that detection is not a problem, as one can just split shares into groups and all should produce the same result.
Reconstruction with corrupted shares is a bit trickier, as to allow t shares to be corrupted, we need at least t^2 shares, so that the number of valid shares dominate. 
Since this is just extra code and higher constraints (e.g. more shares needed to be fetched), we can add this feature latter and currently only request as many shares as needed to reconstruct the key.

Reply:
Note that blame assignment / traitor detection is done via a functor returned with the reconstructed secret, so that the functor's execution can be handed off to a separate thread.  In particular, this means that contract execution is not blocked on the blame assignment computation.

If we don't require significantly more than t+1 shares, we will not be able to detect share corruption.  Share corruption is distinct from blame assignment, and if we are worried about confidentiality loss due to a bad shareholder it seems reasonable to also worry about integrity loss.  Also note that the higher cost is primarily on the _first_ time a key is ever generated/used; subsequent uses we will have additional means of verifying the reconstructed secret that reduces the cost of corrupt share detection -- including reducing latency, since we should not have to wait for all shares to be received before reconstruction.